src/shared_library.c

```diff
@@ -0,0 +1,93 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+#include "rcutils/shared_library.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+
+rcutils_shared_library_t *
+rcutils_get_shared_library(const char * library_path)
+{
+  rcutils_shared_library_t * lib = (rcutils_shared_library_t *)
+    malloc(sizeof(rcutils_shared_library_t));
+
+#ifndef _WIN32
+  lib->lib_pointer = dlopen(library_path, RTLD_LAZY);
+#else
+  lib->lib_pointer = LoadLibrary(library_path);
+#endif  // _WIN32
+  if (!lib->lib_pointer) {
+    free(lib);
+    return NULL;
+  }
+
+  // +1 to accomodate for the null terminator
+  lib->library_path = (char *) malloc((strlen(library_path) + 1) * sizeof(char));
+  snprintf(lib->library_path, strlen(library_path), "%s", library_path);
+
+  return lib;
+}
+
+void *
+rcutils_get_symbol(rcutils_shared_library_t * lib, const char * symbol_name)
+{
+#ifndef _WIN32
+  void * lib_symbol = dlsym(lib->lib_pointer, symbol_name);
+  const char * dlsym_error = dlerror();
+  if (dlsym_error) {
+    fprintf(
+      stderr, "Cannot load symbol '%s' in shared library '%s'",
+      symbol_name, lib->library_path);
+    dlclose(lib->lib_pointer);
+    free(lib->library_path);
+    free(lib);
+    return NULL;
+  }
+#else
+  void * lib_symbol = GetProcAddress(lib->lib_pointer, symbol_name);
+  if (!lib_symbol) {
+    fprintf(
+      stderr, "Cannot load symbol '%s' in shared library '%s'",
+      symbol_name, lib->library_path);
+    FreeLibrary(lib->lib_pointer);
+    free(lib->library_path);
+    free(lib);
+    return NULL;
+  }
+#endif  // _WIN32
+  return lib_symbol;
+}
+
+void
+rcutils_unload_library(rcutils_shared_library_t * lib)
+{
+#ifndef _WIN32
+  dlclose(lib->lib_pointer);
```

Comment:
The return code should be checked and probably be exposed.

Reply:
There is no code error in dlclose()

> The function dlclose() decrements the reference count on the dynamic library handle handle. If the reference count drops to zero and no other loaded libraries use symbols in it, then the dynamic library is unloaded.

---- 

### Windows

> If the function succeeds, the return value is nonzero.
> If the function fails, the return value is zero. To get extended error information, call the GetLastError function.

But I can do something in Windows

Reply:
You might want to also cite the next sentence after paragraph you copy-n-pasted (https://linux.die.net/man/3/dlclose):

> The function dlclose() returns 0 on success, and nonzero on error.

Reply:
According to https://linux.die.net/man/3/dlclose there is a non-zero return code on error with `dlclose()`:

> dlclose()
>
> The function dlclose() decrements the reference count on the dynamic library handle handle. If the reference count drops to zero and no other loaded libraries use symbols in it, then the dynamic library is unloaded.
>
> The function dlclose() returns 0 on success, and nonzero on error. 

And then there is `dlerror` to check the error:

> dlerror()
>
> The function dlerror() returns a human readable string describing the most recent error that occurred from dlopen(), dlsym() or dlclose() since the last call to dlerror(). It returns NULL if no errors have occurred since initialization or since it was last called. 

So I think it should check the return code, set the error state using `RCUTILS_SET_ERROR_MSG(msg)` and then return an `rcutils_ret_t` in the event that a non-zero is returned from `dlclose()`.